tracer/src/Datadog.Trace/Iast/EvidenceConverter.cs

```diff
@@ -95,145 +105,398 @@ private void ToJsonTaintedValue(JsonWriter writer, string value, Range[] ranges)
+        {
+            this.Value = value;
+            this.Source = range.Source;
+
+            // shift ranges to the start of the tainted range and sort them
+            this.SensitiveRanges = new LinkedList<Range>(intersections.Select(r => r.Shift(-range.Start)).OrderBy(r => r.Start));
+            this.IsRedacted = false;
+        }
+
+        public string? Value { get; }
+
+        public Source? Source { get; }
+
+        public bool IsRedacted { get; }
+
+        public LinkedList<Range>? SensitiveRanges { get; }
+
+        public bool ShouldRedact => !IsRedacted && (SensitiveRanges != null || (Source != null && Source.IsSensitive));
+
+        public List<ValuePart> Split()
+        {
+            var parts = new List<ValuePart>();
+            if (Value != null)
+            {
+                if (Source != null && Source.IsSensitive)
                 {
-                    nextTainted.Value.Source?.MarkAsRedacted();
-                    nextSensitive = Poll(sensitive);
+                    // redact the full tainted value as the source is sensitive (password, certificate, ...)
+                    AddValuePart(0, Value.Length, true, parts);
                 }
-
-                // If the tainted range intersects sensitive ranges, we redact the tainted range source and split the sensitive range
-                if (nextSensitive != null && nextSensitive.Value.Intersects(nextTainted.Value))
+                else if (SensitiveRanges != null)
                 {
-                    nextTainted.Value.Source?.MarkAsRedacted();
-                    var rest = nextSensitive.Value.Remove(nextTainted.Value);
-                    nextSensitive = rest.Count > 0 ? rest[0] : null;
+                    // redact only sensitive parts
+                    int index = 0;
+                    foreach (var sensitive in SensitiveRanges)
+                    {
+                        var start = sensitive.Start;
+                        var end = sensitive.Start + sensitive.Length;
+                        // append previous tainted chunk (if any)
+                        AddValuePart(index, start, false, parts);
+                        // append current sensitive tainted chunk
+                        AddValuePart(start, end, true, parts);
+                        index = end;
+                    }
+
+                    // append last tainted chunk (if any)
+                    AddValuePart(index, Value.Length, false, parts);
                 }
+            }
 
-                // We write the range value redacted or not depensing on the source status
-                if (nextTainted.Value.Source != null && nextTainted.Value.Source.IsRedacted())
+            return parts;
+        }
+
+        private void AddValuePart(int start, int end, bool redact, List<ValuePart> valueParts)
+        {
+            if (start < end)
+            {
+                var chunk = Value!.Substring(start, end - start);
```

Comment:
are we sure `Value` cant be null here?

Reply:
The function is always called from scenarios with Value != null, but extra checks won't hurt